doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,191 @@
+
+This is the simplest example when value numbers kind of flow through and are the same, and give us a sense that we are talking about the same value which has the same underling properties.
+
+On the contrary, to give us an intuitive sense for a case when we won't have the same value number is to examine what happens during control flow:
+
+```
+if (...) {
+    x = 1;
+} else {
+    x = 2;
+}
+
+y = x;
+```
+
+In this case the value number of the right hand side of the assignment `y = x` will be different than the value number of the left hand side of the stores to `x` in the if-then-else blocks, because at compile time we may not be able to determine which path we would take, so all three nodes will have different value numbers.
+
+Another important property of value numbers that we should understand is that the notion is beyond just syntax. We can have two expressions such as `x + 5` and `y + 5` where `x` and `y` are known to be equal over a particular range, and thus the two addition expressions will also have the same value number. Thus value numbers are not all just trivial use-def chains and parent-child sort of relations. We can have different nodes (expressions) that don't particularly look alike have the same value number.
+
+This is a very high level summary of what value numbering is. There are excellent resources in literature or online to learn more about value numbering.
+
+## Constraints and Handlers
+
+[Value propagation constraints](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp) are the fundamental classes which are flowed around by the value propagation analysis. Their purpose is the represent constraints determined by the analysis on _value numbers_. If for example we had an expression such as `x = 5` then we would create a constraint which represents the constant value of `5` and we would associate this constraint with the value numbers for the nodes representing the left hand side (`iconst`) and right hand side (`istore`) nodes. Remember that these two nodes would have the same value number because we can prove that they are equal. The VP handlers consume these relations between value numbers and constraints to optimize the program.
+
+[Value propagation handlers](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPHandlers.cpp) are functions which operate on the set of constraints associated with the nodes children and produce constraints for the parent node. There are many handler functions associated with many opcodes. The [value propagation table](https://github.com/eclipse/omr/blob/ad822fbd7f17f1aa1fddafa355cb18fac4247a46/compiler/optimizer/ValuePropagationTable.hpp) describes which function to execute for each particular opcode. Some opcodes are not interesting and all they do is constrain their children, while others perform an interesting operation.
+
```

Comment:
`associated with the nodes children` might be missing an apostrophe.

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/9fb1b0b92e8935d254254edca4bd50d3a6c93e48..d9691be5fa5c4c985b7f2c6c4493e532b222cb40).